designs/disruption-controls-by-reason.md

```diff
@@ -0,0 +1,312 @@
+# Disruption Controls By Reason
+## TOC and Overview
+1. [Overview](#overview)
+2. [User Scenarios](#user-scenarios)
+3. [Known Requirements](#clarifying-the-requirements-and-behavior)
+   - [Q: Default or Undefined Reason Case Handling](#q-default-or-undefined-reason-case-handling)
+   - [Q: Calculation of Allowed Disruptions in a Multi-Reason World](#q-calculation-of-allowed-disruptions-in-a-multi-reason-world)
+   - [Q: Handling an Unspecified Default Reason](#q-handling-an-unspecified-default-reason)
+   - [Q: Definition of Default in Budgets](#q-definition-of-default-in-budgets)
+4. [API Design](#api-design)
+   - [Approach A: Extending the Budget API to Specify a Reason](#approach-a-extending-the-budget-api-to-specify-a-reason)
+      * [List Approach: Multiple Reasons per Budget](#list-approach-multiple-reasons-per-budget)
+      * [Single Reason Approach: One Reason per Budget](#single-reason-approach-one-reason-per-budget)
+      * [Pros and Cons for Both Approaches](#pros-and-cons-for-both-approaches)
+      * [Preferred Option: List Approach](#preferred-option-list-approach)
+   - [Approach B: Defining Per Reason Controls](#approach-b-defining-per-reason-controls)
+      * [Pros and Cons](#pros-and-cons)
+   - [API Design Conclusion: Preferred Design](#api-design-conclusion-preferred-design)
+
+## User Scenarios 
+1. Users need the capability to schedule upgrades only during business hours or within more restricted time windows. Additionally, they require a system that doesn't compromise the cost savings from consolidation when upgrades are blocked due to drift.
```

Comment:
The one question that I have on this scenario: If we are consolidated with a replacement, the replacement node is still going to be on a new version and is going to be effectively using the new configuration, so blocking drift isn't really doing anything here in that sense. We are really just blocking drift if and only if the nodes aren't underutilized enough. Is that enough or even what the user wants? I'm really trying to dig in here to understand why a user would block a node upgrade and if we are really fulfilling their use-case by blocking this node upgrade on some operations but not others.

Reply:
Yeah thats a fair point. Again this flaw pops up where there is some bleed where the user specifies: "don't disrupt due to drifted" but then those nodes will still be liable for other disruption. 

It doesn't succeed very well at blocking NodeImageUpgrade and the more I think about it the more I think that these controls do not meet the upgrade user story in a complete way. 


CC: @charliedmcb  who is cooking up some additional more focused upgrade controls for karpenter.